### ISSUE TITLE
Bug: [require-await] false positives with AsyncDisposableStack

### ISSUE THREAD

**Issue by lifeiscontent** [typescript-eslint/typescript-eslint#11819]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=FAYw9gdgzgLgBLAhiA1nAvHCBTA7nAQSgE8IQARASygAcwpEAjAG2wGUZkUAKASmFbwAJtTpRsQjHABmiZuOBJUAOiHZp2AE7dEJMnD4YAfHADewOHBG16EqTE0BXbMAC%2B-YIlyJK8JSgBtNmIAW0YwZmVdUgpRWwBdPmBQSCgI7CiocU0YbmsxCV4AbmTsAA86HLM4VyKgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aRAR1ko9oEMA7v0r50URNGgB7aJHBgAviEVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const stack = new AsyncDisposableStack()
let disposed = false
stack.defer(async () => {
  disposed = true
})

await stack[Symbol.asyncDispose]()

console.assert(disposed);

export { };
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/require-await": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I'd expect this rule to not trigger on false positives

### Actual Result

triggers on false positives

**Comment by kirkwaiblinger** [comment]
Hi @lifeiscontent! 

How is this a false positive? Can't you supply an ordinary function here, since the implementation doesn't contain any async behavior?

**Comment by bradzacher** [comment]
The type is `defer(onDisposeAsync: () => PromiseLike<void> | void): void` -- so you can pass an ordinary function just fine. So having `async` without an await or returning a promise is correct to report on.
